nltk/tag/mallet.py

```diff
@@ -0,0 +1,249 @@
+                verbose=verbose)
+        
+        self._mallet_dep_jar = find_jar(
+                self._JAR_DEPS, path_to_mallet_deps_jar,
+                env_vars=('MALLET_DEPS',), searchpath=(), url=_mallet_url,
+                verbose=verbose) 
+        
+        self.java_options = java_options
+        self._model_file = ''
+    
+    def set_model_file(self, model_file):
+        self._model_file = model_file
+            
+    def _get_features(self,data):
+        """
+        Extract basic features about this word including 
+             - Current Word 
+             - Is Capitalized ?
+             - Has Punctuation ?
+             - Has Number ?
+             - Suffixes up to length 3 
+        :return : a string which contains the features   
+        
+        >>> from nltk.tag.mallet import MalletTagger
+        >>> mt = MalletTagger()
+        >>> mt._get_features(u'University1')
+        u'WORD_University1 CAPITALIZATION HAS_NUM SUF_1 SUF_y1 SUF_ty1'
+        >>> mt._get_features(u'.?')
+        u'WORD_.? PUNCTUATION SUF_?'
+        """ 
+        feature_list = 'WORD_' + data + ' ' 
+        # Capitalization 
+        if data[0].isupper():
+            feature_list += 'CAPITALIZATION '
+        
+        # Number 
+        pattern = re.compile('\\d')
+        if re.search(pattern, data) is not None:
+            feature_list += 'HAS_NUM ' 
+        
+        # Punctuation
+        punc_cat = set(["Pc", "Pd", "Ps", "Pe", "Pi", "Pf", "Po"])
+        if all (unicodedata.category(x) in punc_cat for x in data):
+            feature_list += 'PUNCTUATION '
+        
+        # Suffix up to length 3
+        if len(data) > 1:
+            feature_list += ('SUF_' + data[-1:] + ' ') 
+        if len(data) > 2: 
+            feature_list += ('SUF_' + data[-2:] + ' ')    
+        if len(data) > 3: 
+            feature_list += ('SUF_' + data[-3:] + ' ')
+        
+        return feature_list.strip()
+        
+    def tag_sents(self, sents):
+        '''
+        Tag a list of sentences. NB before using this function, user should specify the mode_file either by 
+                       - Train a new model using ``train'' function 
+                       - Use the pre-trained model which is set via ``set_model_file'' function  
+        :params sentences : list of sentences needed to tag. 
+        :type sentences : list(list(str))
+        :return : list of tagged sentences. 
+        :rtype : list (list (tuple(str,str))) 
+        '''
+        
+        # We need the list of sentences instead of the list generator for matching the input and output  
+        sentences = list(sents) 
+        
+        # First, build the test file 
+        input_file = tempfile.NamedTemporaryFile(
+                prefix='mallet_tagger.test',
+                dir=tempfile.gettempdir(),
+                delete=False)
+
+        for sent in sentences: 
+            for token in sent:
+                data = unicode(token)
```

Comment:
What is this line for? If the input is unicode then unicode(token) does nothing; if it is not unicode then the text is decoded using some system-depending encoding which is not good. And there is no "unicode" symbol in Python 3.x, so it doesn't work in Python 3.x